staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/controller.go

```diff
@@ -76,13 +77,28 @@ func NewAggregationController(downloader *aggregator.Downloader, openAPIAggregat
 }
 
 // Run starts OpenAPI AggregationController
-func (c *AggregationController) Run(stopCh <-chan struct{}) {
+func (c *AggregationController) Run(stopCh <-chan struct{}, initialAggrDoneCh chan<- struct{}) {
 	defer utilruntime.HandleCrash()
 	defer c.queue.ShutDown()
 
 	klog.Info("Starting OpenAPI AggregationController")
 	defer klog.Info("Shutting down OpenAPI AggregationController")
 
+	// Initially aggregate OpenAPI spec before running worker.
+	var wg sync.WaitGroup
+	for _, name := range c.openAPIAggregationManager.GetAPIServiceNames() {
+		wg.Add(1)
+		go func(name string, wg *sync.WaitGroup) {
+			defer wg.Done()
+			if _, err := c.syncHandler(name); err != nil {
```

Comment:
I can't remember if this method is threadsafe per-key.

Reply:
It's not a good idea, even if it is thread safe. I think it will trigger multiple spec aggregations this way.

Reply:
hmm, actually not so sure. The existing code is hard to follow.

Reply:
GetAPIServiceInfo is thread safe https://github.com/kubernetes/kubernetes/blob/e95983be570fb0c95c8341ecadd33f5577afb9cf/staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/aggregator/aggregator.go#L337

UpdateApiServiceSpec is also thread safe https://github.com/kubernetes/kubernetes/blob/e95983be570fb0c95c8341ecadd33f5577afb9cf/staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/aggregator/aggregator.go#L279.

The only place I'm not sure is Downloader https://github.com/kubernetes/kubernetes/blob/e95983be570fb0c95c8341ecadd33f5577afb9cf/staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/aggregator/downloader.go#L54

Reply:
> It's not a good idea, even if it is thread safe. I think it will trigger multiple spec aggregations this way.

@sttts maybe I can apply the method you use in here https://github.com/kubernetes/kubernetes/blob/e95983be570fb0c95c8341ecadd33f5577afb9cf/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_discovery_controller.go#L217.

I did not use wait.Poll in here not to block apiserver.